File: python/letters.py
#!/usr/bin/env python3

"""
# letters.py

Solve a Countdown letters game.

Copyright 2022 Conway
Licensed under the GNU General Public License v3.0 (GPL-3.0-only).
This is free software with NO WARRANTY etc. etc., see LICENSE.
"""


import argparse


def normalise_letters(string):
  
  return string.strip().upper()


def is_valid(word, input_letters):
  
  return all(
    word.count(letter) <= input_letters.count(letter)
    for letter in word
  )


def compute_valid_word_list(word_list, input_letters):
  
  return [
    word
      for word in word_list
      if is_valid(word, input_letters)
  ]


MAX_RESULTS_DEFAULT = 30
WORD_LIST_FILE_NAME_DEFAULT = '../yawl.txt'


def parse_command_line_arguments():
  
  parser = \
          argparse.ArgumentParser(
            description='Solve a Countdown letters game.'
          )
  
  parser.add_argument(
    'input_letters',
    metavar='LETTERS',
    type=str,
    help='string containing the letters that can be used to form words',
  )
  
  parser.add_argument(
    '-m', dest="max_results_count",
    metavar='MAX_RESULTS',
    type=int,
    default=MAX_RESULTS_DEFAULT,
    help=f'maximum number of output results (default {MAX_RESULTS_DEFAULT})',
  )
  
  parser.add_argument(
    '-w', dest="word_list_file",
    metavar='WORD_LIST',
    type=argparse.FileType('r', encoding='UTF-8'),
    default=WORD_LIST_FILE_NAME_DEFAULT,
    help=f'word list file name (default {WORD_LIST_FILE_NAME_DEFAULT})',
  )
  
  return parser.parse_args()


def print_results(valid_word_list, max_results_count):
  
  for word in valid_word_list[:max_results_count]:
    score = len(word)
    print(f'{score}\t{word}')


def main():
  
  parsed_arguments = parse_command_line_arguments()
  
  input_letters = parsed_arguments.input_letters
  max_results_count = parsed_arguments.max_results_count
  word_list_file = parsed_arguments.word_list_file
  
  input_letters = normalise_letters(input_letters)
  word_list = [
    normalise_letters(word)
      for word in word_list_file.read().splitlines()
  ]
  
  valid_word_list = \
          sorted(
            compute_valid_word_list(word_list, input_letters),
            key=len, reverse=True,
          )
  
  print_results(valid_word_list, max_results_count)


if __name__ == '__main__':
  
  main()

File: java/Letters.java
/*
  # Letters.java
  
  Solve a Countdown letters game.
  
  Copyright 2022 Conway
  Licensed under the GNU General Public License v3.0 (GPL-3.0-only).
  This is free software with NO WARRANTY etc. etc., see LICENSE.
*/

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class Letters
{
  private static final int MAX_RESULTS_DEFAULT = 30;
  private static final String WORD_LIST_FILE_NAME_DEFAULT = "../yawl.txt";
  
  public static String normaliseLetters(final String string)
  {
    return string.strip().toUpperCase();
  }
  
  public static List<String> toCharacterList(final String string)
  {
    return
      string
        .codePoints()
        .mapToObj(codePoint -> new String(new int[] {codePoint}, 0, 1))
        .collect(Collectors.toList());
  }
  
  public static boolean isValid(final String word, final String inputLetters)
  {
    final List<String> wordLetterList = toCharacterList(word);
    final List<String> inputLetterList = toCharacterList(inputLetters);
    
    for (final String letter : wordLetterList)
    {
      if (Collections.frequency(wordLetterList, letter) > Collections.frequency(inputLetterList, letter))
      {
        return false;
      }
    }
    
    return true;
  }
  
  public static List<String> computeValidWordList(final List<String> wordList, final String inputLetters)
  {
    final List<String> validWordList = new ArrayList<>();
    for (final String word : wordList)
    {
      if (isValid(word, inputLetters))
      {
        validWordList.add(word);
      }
    }
    
    return validWordList;
  }
  
  private static Map<String, Object[]> parseCommandLineArguments(final String[] arguments)
  {
    final ArgumentParser argumentParser = new ArgumentParser("Letters", "Solve a Countdown letters game.");
    
    argumentParser.addPositionalArgument(
      "inputLetters", "LETTERS",
      "string containing the letters that can be used to form words",
      1, ArgumentParser.TO_STRING
    );
    
    argumentParser.addOptionalArgument(
      "maxResultsCount", new String[]{"-m"}, "MAX_RESULTS",
      String.format("maximum number of output results (default %d)", MAX_RESULTS_DEFAULT),
      1, new Object[]{MAX_RESULTS_DEFAULT}, ArgumentParser.TO_INTEGER
    );
    
    argumentParser.addOptionalArgument(
      "wordListFile", new String[]{"-w"}, "WORD_LIST",
      String.format("word list file name (default %s)", WORD_LIST_FILE_NAME_DEFAULT),
      1, new Object[]{WORD_LIST_FILE_NAME_DEFAULT}, ArgumentParser.TO_STRING
    );
    
    return argumentParser.parseCommandLineArguments(arguments);
  }
  
  private static void printResults(final List<String> validWordList, final int maxResultsCount)
  {
    final int resultsCount = Math.min(maxResultsCount, validWordList.size());
    for (final String word : validWordList.subList(0, resultsCount))
    {
      final int score = word.codePointCount(0, word.length());
      System.out.printf("%d\t%s%n", score, word);
    }
  }
  
  public static void main(final String[] arguments)
  {
    final Map<String, Object[]> valuesFromName = parseCommandLineArguments(arguments);
    
    final int maxResultsCount = (int) valuesFromName.get("maxResultsCount")[0];
    final String wordListFile = (String) valuesFromName.get("wordListFile")[0];
    String inputLetters = (String) valuesFromName.get("inputLetters")[0];
    
    inputLetters = normaliseLetters(inputLetters);
    
    final List<String> wordList = new ArrayList<>();
    try
    {
      final FileReader fileReader = new FileReader(wordListFile);
      final BufferedReader bufferedReader = new BufferedReader(fileReader);
      
      String line;
      while ((line = bufferedReader.readLine()) != null)
      {
        wordList.add(normaliseLetters(line));
      }
    }
    catch (IOException exception)
    {
      exception.printStackTrace();
    }
    
    final List<String> validWordList = computeValidWordList(wordList, inputLetters);
    validWordList.sort(Comparator.comparingInt(string -> -string.length()));
    
    printResults(validWordList, maxResultsCount);
  }
}
